src/librustc_mir/build/matches/mod.rs

```diff
@@ -688,26 +742,10 @@ impl<'a, 'gcx, 'tcx> Builder<'a, 'gcx, 'tcx> {
             self.cfg.terminate(block, source_info,
                                TerminatorKind::if_(self.hir.tcx(), cond, arm_block, otherwise));
             Some(otherwise)
```

Comment:
You also want to add a false edge (pointing to the imaginary target block) right *after* the guard - that would fix #31287 in the future-proof case (you can also add a test for both the issue and the future-proof to see that MIR borrowck correctly emits an error):
```Rust
let my_str = "hello".to_owned();
match Some(42) {
    Some(_) if { drop(my_str); false } => {} 
    Some(_) => {}
    None => {
        println!("{}", my_str); //~ ERROR [mir] use of moved value
    }
}
```

I mean, technically you could *just* have an edge after the guard, but that would be a hazard if we allow guard to initialize variables:
```Rust
let my_str;
match Some(42) {
    _ if { my_str = String::new(); } => {}
    _ => {
        println!("{}", my_str); //~ ERROR [mir] use of possibly-uninitialized value
    }
}
```